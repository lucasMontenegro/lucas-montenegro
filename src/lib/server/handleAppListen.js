import globals from "lib/utils/globals"
import isProduction from "lib/utils/isProduction"
export default function handleAppListen (PORT) {
  return () => {
    const name = isProduction() ? `cluster worker ${globals.process.pid}` : `dev server`
    globals.console.log(`Node ${name}: listening on port ${PORT}`)
  }
}