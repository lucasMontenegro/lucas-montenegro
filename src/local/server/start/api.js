const express = require("express")
const todo = require("local/todo/router")
const api = module.exports = express.Router()
api.use(`/todo`, todo)