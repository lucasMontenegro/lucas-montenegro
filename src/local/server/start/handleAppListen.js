const { process, console } = global
module.exports = function handleAppListen (PORT) {
  return () => {
    const name = (
      process.env.NODE_ENV === `production` ? `cluster worker ${process.pid}` : `dev server`
    )
    console.log(`Node ${name}: listening on port ${PORT}`)
  }
}