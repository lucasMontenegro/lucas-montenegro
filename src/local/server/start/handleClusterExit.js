const { console } = global
module.exports = function handleClusterExit (worker, code, signal) {
  const { pid } = worker.process
  console.log(`Node cluster worker ${pid} exited: code ${code}, signal ${signal}`)
}