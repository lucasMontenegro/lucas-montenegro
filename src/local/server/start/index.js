const cluster = require("cluster")
const os = require("os")
const startHttp = require("local/server/start/startHttp")
const handleClusterExit = require("local/server/start/handleClusterExit")
const { console, process } = global
module.exports = function start () {
  // Multi-process to utilize all CPU cores.
  if (process.env.NODE_ENV === `production` && cluster.isMaster) {
    console.log(`Node cluster master ${process.pid} is running`)
    // Fork workers.
    for (let i = 0, n = os.cpus().length; i < n; i++) {
      cluster.fork()
    }
    cluster.on(`exit`, handleClusterExit)
  } else {
    startHttp()
  }
}