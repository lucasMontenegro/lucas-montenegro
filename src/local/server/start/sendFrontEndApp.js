const path = require("path")
const str = path.resolve(process.env.BUILD_PATH, `index.html`)
module.exports = function sendFrontEndApp (req, res) {
  res.sendFile(str)
}