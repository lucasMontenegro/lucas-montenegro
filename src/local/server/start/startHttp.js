const express = require("express")
const api = require("local/server/start/api")
const sendFrontEndApp = require("local/server/start/sendFrontEndApp")
const handleAppListen = require("local/server/start/handleAppListen")
const { process } = global
module.exports = function startHttp () {
  const app = express()
  // Priority serve any static files.
  app.use(express.static(process.env.BUILD_PATH))
  // Answer API requests.
  app.use(`/api`, api())
  // All remaining requests return the React app, so it can handle routing.
  app.get(`*`, sendFrontEndApp)
  const PORT = process.env.PORT || 5000
  app.listen(PORT, handleAppListen(PORT))
}