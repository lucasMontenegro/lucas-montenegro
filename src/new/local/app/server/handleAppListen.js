const { default: globals } = require("new/local/utils/globals")
const { default: isProduction } = require("new/local/utils/isProduction")
exports.default = function handleAppListen (PORT) {
  return () => {
    const name = isProduction() ? `cluster worker ${globals.process.pid}` : `dev server`
    globals.console.log(`Node ${name}: listening on port ${PORT}`)
  }
}