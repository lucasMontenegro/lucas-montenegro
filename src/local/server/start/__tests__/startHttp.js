const mockFn = jest.fn()
const makeFn = name => (...args) => {
  const arr = [name, args]
  mockFn(...arr)
  return arr
}
jest.mock(`express`, () => jest.fn())
const express = require("express")
express.mockImplementation((...args) => {
  mockFn(`express`, args)
  return [`use`, `get`, `listen`].reduce((app, name) => {
    app[name] = makeFn(`app.${name}`)
    return app
  }, {})
})
express.static = makeFn(`express.static`)
jest.mock(`local/server/start/api`, () => `local/server/start/api`)
jest.mock(`local/server/start/sendFrontEndApp`, () => `local/server/start/sendFrontEndApp`)
jest.mock(`local/server/start/handleAppListen`, () => jest.fn())
require("local/server/start/handleAppListen").mockImplementation(makeFn(`handleAppListen`))
const mockGlobals = require("local/utils/mockGlobals")
const globals = { process: { env: { BUILD_PATH: `'process.env.BUILD_PATH'` } } }
const oldGlobals = mockGlobals(globals)
const startHttp = require("local/server/start/startHttp")
mockGlobals(oldGlobals)
describe(`local/server/start/startHttp`, () => {
  afterEach(() => mockFn.mockClear())
  test.each([
    [80],
    [null],
  ])(`should configure and start the http app (PORT %j)`, PORT => {
    globals.process.env.PORT = PORT
    startHttp()
    expect(mockFn.mock.calls).toMatchSnapshot()
  })
})