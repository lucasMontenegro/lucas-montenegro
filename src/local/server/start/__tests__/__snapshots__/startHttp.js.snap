// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`local/server/start/startHttp should configure and start the http app (PORT 80) 1`] = `
Array [
  Array [
    "express",
    Array [],
  ],
  Array [
    "express.static",
    Array [
      "'process.env.BUILD_PATH'",
    ],
  ],
  Array [
    "app.use",
    Array [
      Array [
        "express.static",
        Array [
          "'process.env.BUILD_PATH'",
        ],
      ],
    ],
  ],
  Array [
    "app.use",
    Array [
      "/api",
      "local/server/start/api",
    ],
  ],
  Array [
    "app.get",
    Array [
      "*",
      "local/server/start/sendFrontEndApp",
    ],
  ],
  Array [
    "handleAppListen",
    Array [
      80,
    ],
  ],
  Array [
    "app.listen",
    Array [
      80,
      Array [
        "handleAppListen",
        Array [
          80,
        ],
      ],
    ],
  ],
]
`;

exports[`local/server/start/startHttp should configure and start the http app (PORT null) 1`] = `
Array [
  Array [
    "express",
    Array [],
  ],
  Array [
    "express.static",
    Array [
      "'process.env.BUILD_PATH'",
    ],
  ],
  Array [
    "app.use",
    Array [
      Array [
        "express.static",
        Array [
          "'process.env.BUILD_PATH'",
        ],
      ],
    ],
  ],
  Array [
    "app.use",
    Array [
      "/api",
      "local/server/start/api",
    ],
  ],
  Array [
    "app.get",
    Array [
      "*",
      "local/server/start/sendFrontEndApp",
    ],
  ],
  Array [
    "handleAppListen",
    Array [
      5000,
    ],
  ],
  Array [
    "app.listen",
    Array [
      5000,
      Array [
        "handleAppListen",
        Array [
          5000,
        ],
      ],
    ],
  ],
]
`;
